### PR TITLE
New models and model loading, depending on sklearn version

### PR DESCRIPTION
**Pull request by bdewilde**
Changes:

- Add `sklearn_path` to `compat` module, use it for saving and loading different content extraction models for "old" and "new" versions of `scikit-learn`
- Wrap default content extraction model loading in a try/except; in case of failure, display a warning message and set models to null; this lets devs use other parts of the package even when the default models can't be loaded
- Added newly-trained models for content and content+comments extraction
- _Commented out_ loading and testing of non-default models, but can totally add these back in (or totally remove them) if you'd prefer
- Updated the docstring on the main API method, `ContentExtractionModel.analyze()`

Performance in terms of accuracy/precision/recall/F1 of the new models was on par or slightly better than previous benchmarks. See Issue #33 for details.

### CODE REVIEW

**Comment by bdewilde** [comment]
Looks like the test I modified isn't passing in Travis, but it definitely passes locally!

**Comment by matt-peters** [comment]
Nice!  This looks great and is a huge improvement to maintaining with different sklearn versions.  What version of sklearn do you have locally?  I wonder if the issue with the Travis build is due to different sklearn versions.  The different models will extract slightly different content so we may need one test fixture to support each version of sklearn.

**Comment by bdewilde** [comment]
Hm, that's a possibility... I used sklearn 0.17.1 and 0.18.1 for the "old" and "new" models, respectively. I didn't use, say, 0.16.1 because the API was compatible with 0.17.1 ... I realize now that I didn't update the version requirements in `setup.py`, which should probably be `'scikit-learn>=0.15.2,<=0.19'`. Maintaining separate tests for (all?) different sklearn versions could get ugly, I wonder if there's a way to do approximate equality checks?

**Comment by matt-peters** [comment]
Ah I see.  So your `sklearn_0.15.2_0.17.1` picked model was trained with version 0.17.1 and there is another API change between 0.16.X and 0.17?  In that case, it isn't compatible with 0.15-0.16 and we three need different pickled models to support 0.16 - 0.18.  Ugh.

As for test cases, we could compute the precision, recall, f1 for the predicted version vs the gold standard and make sure they exceed some high threshold like 0.99.

**Comment by bdewilde** [comment]
Ah, maybe I wasn't clear... sklearn's API seemed to be compatible from 0.15.2 to 0.17.1, then changed at 0.18.0 +. So we should only need the two models (for now).

**Comment by bdewilde** [comment]
Hey @matt-peters , you were right: differences in `scikit-learn` versions account for the failed test. Travis uses 0.16.1 (despite changes to `setup.py`) to test the model, while I used 0.17.1 to train it, and although the `ExtraTreesClassifier` API didn't change across versions, the model results changed _just slightly_. I updated the model tests to use the bag-of-words f1 score between predicted and actual content, with a minimum acceptable value of 0.98. Tests pass on all versions: 0.16.1, 0.17.1, and 0.18.1.

File: dragnet/model_training.py [hunk 1]
```diff
@@ -93,8 +94,18 @@ def make_features_from_data(self, data, model, training_or_test='training',
...
-            features = np.vstack((features, training_features))
             this_labels, this_weight, this_tokens = self._get_labels(content, comments)
+
+            if (training_features.shape[0] != len(this_weight) or
+                    len(this_labels) != len(this_weight)):
+                print('\nskipping file because of array shape mismatch...')
+                print('len(blocks) =', len(blocks))
+                print('len(this_labels) =', len(this_labels))
+                print('len(this_weight) =', len(this_weight))
+                # raise ValueError("Number of features, labels and weights do not match!")
```
Comment (matt-peters) [inline dragnet/model_training.py]: Let's continue to raise this an error.  By trapping it, we could mask another larger significant problem and create a much smaller or corrupted data set.  Someone who is retraining without being familiar with the code details is likely to not notice it, so let's make it explicit when it is ignored.

File: dragnet/models.py [hunk 2]
```diff
@@ -1,62 +1,70 @@
...
+    warnings.warn(
+        "Unable to unpickle ContentExtractionModel! "
+        "Your version of scikit-learn ({}) may not be compatible. "
+        "Setting extractors to None.".format(sklearn_version),
+        UserWarning)
+    content_extractor = None
+    content_comments_extractor = None
+    content_and_content_comments_extractor = None
+
+# weninger_model = Weninger()
```
Comment (matt-peters) [inline dragnet/models.py]: You can remove these lines too.

File: test/test_models.py [hunk 3]
```diff
@@ -16,27 +18,32 @@ def setUp(self):
             FIXTURES, 'models_testing.html'), 'r').read()
 
     def test_models(self):
-        models = [kohlschuetter_model,
-                  weninger_model,
-                  kohlschuetter_weninger_model,
-                  kohlschuetter_css_model,
-                  kohlschuetter_css_weninger_model,
-                  content_extractor,
+        # models = [kohlschuetter_model,
```
Comment (matt-peters) [inline test/test_models.py]: These commented out lines can go

**Comment by matt-peters** [comment]
Would you mind also adding your training script to train/evaluate end-to-end?  It will make it easier for someone else to add support for a new sklean version.

Otherwise it looks good except for the few minor comments above.

**Comment by bdewilde** [comment]
Okay, that last commit does everything you mentioned in the comments. Two things:

1. A test in `test_models.py` failed on Travis (ugh), _but_ I could only replicate this behavior intermittently. I believe this can be attributed to the nondeterminism in one of the features. So, I'm going to re-bump up the number of trails from 5 to 10.
2. What should be done about the models in the `pickled_models` directory (not the new ones I've added in sklearn-specific folders)?

**Comment by matt-peters** [comment]
> What should be done about the models in the `pickled_models` directory (not the new ones I've added in sklearn-specific folders)?

I don't see any reason to keep them so let's just delete them.

**Comment by matt-peters** [comment]
👍 Thanks for all your work on this!

**Comment by bdewilde** [comment]
My pleasure! I don't suppose you could push this as a new release to pypi? I'd love to use it in production. :)

**Comment by sistawendy** [comment]
The installation is still failing in Travis like this:

```
creating build/lib.linux-x86_64-2.7/dragnet/pickled_models
error: can't copy 'dragnet/pickled_models/sklearn_0.18.0': doesn't exist or not a regular file
```

**Comment by matt-peters** [comment]
@sistawendy The travis build looks fine to me:  https://travis-ci.org/seomoz/dragnet  ?  Where do you see the failure?

**Comment by sistawendy** [comment]
In a Freshscape build. It's here, if you have access:
https://travis-ci.com/seomoz/freshscape/builds/37887192

**Comment by matt-peters** [comment]
This is fixed in #39 

I will push a new release to pypi tomorrow.

**Comment by bdewilde** [comment]
Hey @matt-peters , did you push a new release? The latest I see on pypi is 1.0.2, which I believe came before this set of changes.

**Comment by matt-peters** [comment]
Just pushed 1.1.0 to do pypi.
